cloudbridge/cloud/providers/gce/services.py

```diff
@@ -654,28 +708,148 @@ def create_floating_ip(self, region=None):
+                                    region=region,
+                                    body={'name': ip_name})
+                            .execute())
             if 'error' in response:
                 return None
             self.provider.wait_for_operation(response, region=region)
             ips = self.floating_ips()
             for ip in ips:
-                if ip.id == response["targetId"]:
+                if ip.id == response['targetId']:
                     return ip
         except:
             return None
 
-    def routers(self):
-        raise NotImplementedError('To be implemented')
+    def routers(self, region=None):
+        if not region:
+            region = self.provider.region_name
+        try:
+            routers = []
+            token = None
+            while True:
+                response = (self.provider
+                                .gce_compute
+                                .routers()
+                                .list(project=self.provider.project_name,
+                                      region=region,
+                                      pageToken=token)
+                                .execute())
+                for router in response.get('items', []):
+                    routers.append(GCERouter(self.provider, router))
+                if 'nextPageToken' not in response:
+                    break
+                token = response['nextPageToken']
+            return routers
+        except:
+            return []
+
+    def create_router(self, name=None, network=None, region=None):
+        network_url = 'global/networks/default'
+        if isinstance(network, GCENetwork):
+            network_url = network.resource_url
+        if not region:
+            region = self.provider.region_name
+        try:
+            response = (self.provider
+                            .gce_compute
+                            .routers()
+                            .insert(project=self.provider.project_name,
+                                    region=region,
+                                    body={'name': name,
+                                          'network': network_url})
+                            .execute())
+            if 'error' in response:
+                return None
+            self.provider.wait_for_operation(response, region=region)
+            routers = self.routers()
+            for router in routers:
+                if router.id == response['targetId']:
+                    return router
+        except:
+            return None
+
+
+class GCESubnetService(BaseSubnetService):
+
+    def __init__(self, provider):
+        super(GCESubnetService, self).__init__(provider)
 
-    def create_router(self, name=None):
+    def get(self, subnet_id):
+        for subnet in self.list():
+            if subnet.id == subnet_id:
+                return subnet
+        return None
+
+    def list(self, network=None, region=None):
```

Comment:
@chiniforooshan @afgane The list method needs the marker and limit parameters as per the interface. The aws and openstack implementations also have the same bug here, because they don't accept limit and marker. We need an accompanying test which detects this. With limit and marker, we can also simplify the implementation below because there's no need to iterate through all the results.